### PR TITLE
CMake: new parser

### PR DESCRIPTION
**Pull request by hadrielk**
As mentioned in issue #1816, this is a CMake file parser using the optlib multitable regex feature.

### CODE REVIEW

File: win32/ctags_vs2013.vcxproj.filters [hunk 1]
```diff
@@ -660,4 +663,4 @@
       <Filter>Header Files</Filter>
     </ClInclude>
   </ItemGroup>
-</Project>
\ No newline at end of file
+</Project>
```
Comment (masatake) [inline win32/ctags_vs2013.vcxproj.filters]: This change for the end of file should not be added. git add -p may help you.

**Comment by masatake** [comment]
Other than one minor point about the build file for windoes. everything looks good to me.
I made one pull request to you to support end field. My pull request supports scope, too. However, I'm not sure the scope filed makes sense in cmake syntax. If my pull request is acceptable, could you merge and push the changes here?
(https://github.com/hadrielk/ctags/pull/1)

After visiting your profile page I know why you needs cmake parser; you are a wireshark hacker.
Uuniversal-ctags can offer more attractions to you, visit https://github.com/universal-ctags/ctags/issues/1798 :-P

**Comment by masatake** [comment]
( I run "make chop LANGUAGES=CMake", "make slap LANGUAGES=CMake", and "make fuzz LANGUAGES=CMake". All are passed.)

**Comment by masatake** [comment]
The techniques used in the cmake.ctags are impressive. I have to consider applying them to my PuppetManifest parser.

File: win32/ctags_vs2013.vcxproj.filters (hunk 1 above)
Comment (hadrielk) [reply win32/ctags_vs2013.vcxproj.filters]: Oops. That's my editor being too smart for its own good. I'll fix it.

**Comment by hadrielk** [comment]
> I'm not sure the scope filed makes sense in cmake syntax

There's no namespace concept in CMake - they do have a concept of "scope", but not in the sense that ctags would care I think (other than the `end:` field). Most of their scoping is based on directory information. Except variables (created with `set()`), which if set in a function/macro are in that function's scope only, unless they specify the `PARENT_SCOPE` argument.

Do you think I should try to add the scope for variable tags within functions/macros? (ie, use `{scope=ref}`) If so I'll have to also look for `PARENT_SCOPE` and if it's there then not set the scope reference.

File: optlib/cmake.ctags [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+# capture a required trailing '[# \t\n\)]', to verify the full name token matched
+# the name's pattern, but then we advanceTo=2start for the next round, so that
+# we don't go past a potential '#' comment token but instead match it again in
+# the main table as a comment (or whitespace if it was whitespace).
+#
+
+#
+# variable
+#
+--_mtable-regex-CMake=variable/([A-Za-z0-9_.-]+)([# \t\n\)])/\1/v/{tleave}{advanceTo=2start}
```
Comment (hadrielk) [inline optlib/cmake.ctags]: bug: all these `{advanceTo=2start}` should be `{_advanceTo=2start}`

**Comment by masatake** [comment]
Could you squash 08c4658  to acdd8 ? 
Could you squash the changes for optlib/cmake.ctags in 66c62e0 to 729059f  ? 
Could you squash the changes for  Units/parser-cmake.r/cmake-comments.d/expected.tags and  Units/parser-cmake.r/cmake-comments.d/input.ctags in 66c62e0 to d3803b4 ?
Could you sqash the changes for optlib/cmake.c in  66c62e0 to acdd8 ?

**Comment by hadrielk** [comment]
Would you mind if I just squashed all my changes down to one commit? It would be easier.

**Comment by masatake** [comment]
Merging it one ("CMake: new parser")  is o.k.

**Comment by hadrielk** [comment]
Done.

**Comment by masatake** [comment]
Thank you!

**Pull request merged**